qiskit/primitives/containers/sampler_pub.py

```diff
@@ -0,0 +1,139 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+
+"""
+Sampler Pub class
+"""
+
+from __future__ import annotations
+
+from typing import Tuple, Union
+from numbers import Integral
+
+from qiskit import QuantumCircuit
+
+from .bindings_array import BindingsArray, BindingsArrayLike
+from .shape import ShapedMixin
+
+
+class SamplerPub(ShapedMixin):
+    """Pub (Primitive Unified Bloc) for Sampler.
+
+    Pub is composed of tuple (circuit, parameter_values, shots).
+
+    If shots are provided this number of shots will be run with the sampler,
+    if ``shots=None`` the number of run shots is determined by the sampler.
+    """
+
+    __slots__ = ("_circuit", "_parameter_values", "_shots")
+
+    def __init__(
+        self,
+        circuit: QuantumCircuit,
+        parameter_values: BindingsArray | None = None,
+        shots: int | None = None,
```

Comment:
Same question as EstimatorV2, why is `shots` part of `__init__` (and `run()`) when other execution related options are in options?

Reply:
Options are not part of the interface since they dont place any meaningful constraints on implementations, so cannot be generically assumed in algorithms or middle ware built on top of primitives. Primitive implementation is free to do whatever it likes with options. The only thing the interface requires is shots, since a Sampler has to return that many samples. We removed it from init so now it is only in pubs/run.

Reply:
Well if `shots` is a required option, I'd say it makes `options` meaningful. Furthermore, I'd argue the primary reason for having a base class is to provide a consistent user experience, even if it makes less sense from development point of view. Having a standard way to set/query options was specifically called out while reviewing the options RFC.